api/src/main/java/org/apache/iceberg/RewriteManifests.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+/**
+ * API for rewriting manifests for a table.
+ * <p>
+ * This API accumulates manifest files, produces a new {@link Snapshot} of the table
+ * described only by the manifest files that were added, and commits that snapshot as the
+ * current.
+ * <p>
+ * When committing, these changes will be applied to the latest table snapshot. Commit conflicts
+ * will be resolved by applying the changes to the new latest snapshot and reattempting the commit.
+ */
+public interface RewriteManifests extends SnapshotUpdate<RewriteManifests> {
+
+  /**
+   * Append a {@link DataFile} to the table, and specify the manifest that should contain it with
+   * the given key. All files with the same key will be written to the same manifest.
+   *
+   * @param file a data file
+   * @param key  a key specifying the manifest
+   * @return this for method chaining
+   */
+  RewriteManifests appendFile(DataFile file, Object key);
```

Comment:
This appears to be a different kind of append, not just a metadata rewrite. I would expect `RewriteManifests` to operate like `RewriteFiles` but on manifests instead of data files. That way you can rewrite the manifests as you choose.

Reply:
Actually, I see why you probably wanted to do it this way, because the manifest writer needs to write the ID of the snapshot where it is added. And we need to keep track of the snapshot where each file was added.

Reply:
I was hoping to allow rewriting only part of the metadata, so keep some existing manifest files (e.g. those that were already rewritten) and replace others with new files. I'll give that some more thought...

Reply:
I implemented what you suggested, it makes the API much easier to use I think. I also added filter() to include only manifests to process (and keep the rest as-is).